drivers: sensor: Add STM32 VREF+ sensor

knthm opened the pull request:
### Summary

This PR adds a driver for measuring the `VREF+` voltage on supported STM32 SoCs, by calculating it from data measured via the ADC's internal `VREFINT` channel, and its factory calibration data.

In many package configurations, `VREF+` is bonded to `VDDA`, allowing direct measurement of `VDDA` in device configurations where it might fluctuate, e.g. lithium CR-series battery powered devices without external voltage regulation.

I've only physically tested this on a STM32WLE5CCU6-based device, but the functionality provided is identical in the other SoC series for which I've added the `vref` node.

### Change Description

Add VREF+ sensor driver.
Add vref DT node definition for setting `VREFINT` calibration data parameters.
Add vref DT node and calibration parameters to supported STM32 SoC series DTS.

### Impact and Future Work

The logic in this driver could be used as a basis to implement supply voltage independent ADC measurements, as the STM32 ADC driver currently only supports measurements against the static `st,stm32-adc` `vref-mv` property. For more details see the appropriate chapter in the reference manual below.

### References

See RM0461, chapter 16.9 - Temperature sensor and internal reference voltage.

1. [RM0461 Reference manual - STM32WLEx advanced Arm-based 32-bit MCUs with sub-GHz radio solution](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf)

knthm committed:
drivers: sensor: Add STM32 VREF+ sensor

Add VREF+ sensor driver and DT node definition.

This driver allows determining the actual voltage applied to an SoC's
VREF+ pin, by comparing the VREFINT internal bandgap voltage reference
with its factory calibration data.

In packages where VREF+ is bonded to VDDA, this permits direct measurement
of VDDA voltage.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
dts: arm: st: Add vref node to supported STM32 SoCs

Add new vref node to the DTS definitions of supported SoCs.
Extend DTS ADC channel properties where missing.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

KloudJack commented:
- Split PR into multiple commits based on components affected

KloudJack commented:
- Split PR into multiple commits based on components affected
- Amend DTS commit with commit message body

KloudJack commented:
- Split PR into multiple commits based on components affected
- Amend DTS commit with commit message body
- Amend commit to retrigger Github actions failed due to current [incident](https://www.githubstatus.com/incidents/pr3498h3qkfy)

erwango reviewed (changes requested):
Thanks for this addition.
Looks good to me, but would you mind adding a sample, for zephyr users first, but mostly to enable building and testing in CI ?

FRASTM commented:
> Thanks for this addition. Looks good to me, but would you mind adding a sample, for zephyr users first, but mostly to enable building and testing in CI ?

Yes, a simple sample would help (quite similar to the samples/sensor/stm32_vbat_sensor). On most of the stm32 target boards, the VREF is VDD, so testcase should simply compare ADC vref sensor value to the ADC_REF_INTERNAL.

MaureenHelm commented on drivers/sensor/stm32_vref/stm32_vref.c:
```diff
@@ -0,0 +1,140 @@
+#endif /* CONFIG_SOC_SERIES_STM32H5X */
+
+	/* Calculate VREF+ using VREFINT bandgap voltage and calibration data */
+	vref = cfg->cal_mv * (*cfg->cal_addr) / data->raw;
```
You may want to guard against dividing by zero

KloudJack replied:
Good call, will amend the commit to add this. Thanks!

KloudJack commented:
> Looks good to me, but would you mind adding a sample, for zephyr users first, but mostly to enable building and testing in CI ?

Sounds good, I'll add a sample based on `stm32_vbat`.

> On most of the stm32 target boards, the VREF is VDD, so testcase should simply compare ADC vref sensor value to the ADC_REF_INTERNAL.

That's a bit tricky, since the sensor value will always have some degree of variance.
Matching `ADC_REF_INTERNAL = 3000` when the ADC is reporting `2987 mV` is a bit tricky with the limited regex capabilities of `harness_config`. Or is there another method for comparison that I'm not seeing?

knthm committed:
samples: sensor: Add generic voltage driver example

Add basic sample to demonstrate voltage driver measurements.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

KloudJack commented:
- Rebase on main
- Add guard condition to protect against `data->raw == 0`
- Add sample demonstrating use of STM32 vref driver with basic test harness and board overlays for one of every supported STM32 series SoC

MaureenHelm reviewed (changes requested):
Please make the sample generic so it can support other voltage sensor drivers. See `samples/sensor/die_temp_polling` for example

knthm commented:
I like that idea!

After a brief look at that sample, that'd entail adding a board DT alias for the various voltage sensors an SoC might have. Something like `volt-sensor` might be a concise name?

I'd even go a step further and consolidate the redundancies this would create in the `stm32_vbat` sample and corresponding driver/SoC DTS. That'd be adding the `stm32_vbat` DT node into the supported SoC's DTS (with `status = "disabled"`) and adding those nodes under the new `volt-sensor` alias for the corresponding boards.

Not sure if that'd be better off in its own PR though, any thoughts?

knthm committed:
boards: arm: Add vref node and alias for STM32 SoCs

Add enabled vref node for all boards of supported STM32 SoC series which
have the required adc node enabled.

Add volt-sensor alias pointing to said vref node.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
samples: sensor: Remove redundant stm32_vbat_sensor

Remove stm32_vbat_sensor sample which has been made redundant by the new
generic soc_voltage sample.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
It took some brushing up on my perl-foo, but I've added and enabled the `vref` node with a `volt-sensor0` alias for all boards utilizing a STM32 SoC that supports calibrated `VREF+` measurements—provided that the board has an enabled `adc` node.

Like the `die_temp` node, the `vref` node is now enabled on all the STM32 series boards supporting its functionality by default. I've chosen this approach as it enables easier usage of `VREF+` measurements later on, should the `adc_stm32` driver be expanded to support measurements relative to `VREF+`, instead of the hard-coded DTS `vref-mv` property value.

The sample is now called `soc_voltage` and supports any voltage measurement sensor, provided it is defined under the `volt-sensorX` alias.

Finally, I've migrated and adapted the board overlays from the `stm32_vbat_sensor` sample, and removed this sample as its function is covered entirely by the generic `soc_voltage` sample.

I've tested the sample with `twister --device-testing` on my `olimex_lora_stm32wl_devkit` board (with some out-of-tree DTS to enable missing ADC support), and confirmed it working:

```
[...]
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 errored, 0 skipped with 0 warnings in 12.92 seconds
[...]
```
```
*** Booting Zephyr OS build zephyr-v3.3.0-3905-g526e1426cd68 ***
Sensor voltage[vref]: 3.01 V
Sensor voltage[vbat]: 3.33 V
```

knthm commented:
- Remove `vref` node from STM32H7 boards where `adc3` isn't enabled

knthm commented:
- No change amend to retrigger actions after erroneous workflow failure

erwango commented on samples/sensor/soc_voltage/boards/nucleo_g071rb.overlay:
```diff
@@ -0,0 +1,12 @@
+/ {
```
While you're at, let's move all these to `boards/`

knthm replied:
OK, I'll add the `vbat` node to the appropriate SoCs to compliment this as well.

gautierg-st commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -862,6 +862,8 @@
 			status = "disabled";
 			#io-channel-cells = <1>;
+			temp-channel = <18>;
+			vref-channel = <19>;
```
temp-channel and vref-channel are on different ADCs on different channels across the H7 family.
For H72x/73x, they are on ADC3 ch. 17 & 18.
For H74x/75x, they are on ADC3 ch. 18 & 19.
For H7Ax/7Bx, they are on ADC2 ch. 18 & 19.

If you take a look at stm32h723.dtsi, stm32h743.dtsi, stm32h745.dtsi, stm32h750.dtsi, or stm32h7a3.dtsi, they are all already defined.

But, there is no ADC3 on H7Ax/H7Bx, and ADC3 is already redefined for H72x. So I think we can keep these two lines in this file, but in that case please remove redundant definitions in stm32h743.dtsi, stm32h745.dtsi, stm32h750.dtsi.

I will take care to remove adc3 node in stm32h7a3.dtsi (I already have an open PR on ADC cleanup that I'll use).

knthm replied:
Appreciate the in-depth review!

After a closer look at the HAL and respective datasheets/RMs, I've also found that STM32H7A3 uses a different address for the `VREFINT` calibration data as well. Again, I'd falsely assumed all the STMH7 series shared the same parameters between them.

I've removed the above ADC channel properties from `stm32h7.dtsi`, as well as the `vref` node `io-channels` property, and added them to the individual SoC definitions, while also keeping their individual ADC node channel properties.

It feels a bit clearer to not have property defaults if these get overridden more than once for multiple diverging SoC specifications, even if it is a tad more verbose. IMO the best solution would be an intermediate common DTSI for the H72x/73x, H74x/75x, and H7Ax/7Bx sub-series, but that should probably be a separate PR.

gautierg-st replied:
Good catch about the address for H7A, I hadn't checked that.

erwango reviewed (commented):
Great work, one non blocking suggestion

gautierg-st commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -1032,6 +1034,14 @@
+		compatible = "st,stm32-vref";
+		vrefint-cal-addr = <0x1FF1E860>;
+		vrefint-cal-mv = <3300>;
+		io-channels = <&adc3 19>;
```
To continue on my previous comment, this specific property should be redefined in stm32h723.dtsi and stm32h7a3.dtsi, to select the correct ADC and channel.

knthm replied:
Please see my previous comment.

gautierg-st reviewed (changes requested):
Just a few changes needed around some ADCs quirks, otherwise LGTM

MaureenHelm commented on drivers/sensor/stm32_vref/stm32_vref.c:
```diff
@@ -0,0 +1,145 @@
+	vref /= 1000;
+
+#if defined(CONFIG_SOC_SERIES_STM32H5X)
+	LL_ICACHE_Enable();
```
Is the cache ever disabled before entering the driver?

knthm replied:
Concerning code running on the STM32H5 series, AFAICT the only other places are in the `hwinfo` driver and `stm32_temp` sensor driver, which enable it after executing the critical section affecting ICACHE. It seems to get enabled in `stm32h5_init`.

I'd added this based on the same line in the temp driver, and that can't run simultaneously with the vref sensor, as they both use the same ADC.

The `hwinfo` driver has identical logic added for the same reason: 8d5ff8cb68673eaa0f1023af37148c37ed782bf6. I'm not sure if that could conflict with the ADC-based drivers, as I'm not familiar with the ICACHE/HAL specifics.

MaureenHelm replied:
@erwango are you ok with unconditionally enabling the cache here for STM32H5?

erwango replied:
> @erwango are you ok with unconditionally enabling the cache here for STM32H5?

Yes, It's enabled by default on soc.c file.

MaureenHelm commented on drivers/sensor/stm32_vref/Kconfig:
```diff
@@ -0,0 +1,12 @@
+	bool "STM32 VREF Sensor"
+	default y
+	depends on DT_HAS_ST_STM32_VREF_ENABLED
+	depends on (ADC && SOC_FAMILY_STM32)
```
```suggestion
	select ADC
```

knthm replied:
This is more concise and seems to be in line with the structure of most other driver Kconfigs.

Amended original commit, thanks!

MaureenHelm reviewed (commented):
This looks great, thank you. Just one nit and a question.

knthm committed:
dts: arm: st: Add vbat node to supported STM32 SoCs

Add vbat node to DTS definitions of supported SoCs.
Extend/fix ADC channel properties where missing.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented on dts/arm/st/f3/stm32f3.dtsi:
```diff
@@ -445,6 +445,14 @@
+		compatible = "st,stm32-vref";
+		vrefint-cal-addr = <0x1FFFF7BA>;
+		vrefint-cal-mv = <3300>;
+		io-channels = <&adc1 18>;
```
Added overriding node property `io-channel`  in `stm32f373.dtsi`, thanks! I had only checked the STM32F302 datasheet and Reference Manual and falsely assumed them identical across the STM32F3 series.

knthm commented:
- Rebase on main
- Fix STM32 series ADC/sensor channel DTS definition discrepancies
- Add `vbat` node and ADC channel property to all supported SoC's DTS
- Add `vbat` node to all supported boards which have required ADC enabled

knthm committed:
boards: arm: Add vbat node and alias for STM32 SoCs

Add enabled vbat node for all boards of supported STM32 SoC series which
have the required adc node enabled.

Add volt-sensor alias pointing to said vref node.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
- Add `vbat` node only for boards with SoC supporting the corresponding ADC channel (used board list for `vref` supporting boards initially, `vref` and `vbat` channel support aren't mutually inclusive)
- Enable only either `adc1` or `adc4` `vref` and `vbat` nodes on STM32U5-based boards

knthm commented:
- Add `vbat` node only for boards with SoC supporting the corresponding ADC channel (used board list for `vref` supporting boards initially, `vref` and `vbat` channel support aren't mutually inclusive)
- Enable only either `adc1` or `adc4` `vref` and `vbat` nodes on STM32U5-based boards

knthm commented:
- Remove some `vref` and `vbat` nodes from boards with appropriate ADC not enabled
- Add aliases for board with aliases separate from common dtsi `vref`/`vbat` node definitions

GeorgeCGV commented on drivers/sensor/stm32_vref/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+zephyr_library()
+
+zephyr_library_sources(stm32_vref.c)
```
Shouldn't it be guarded?

```suggestion
zephyr_library_sources_ifdef(CONFIG_STM32_VREF, stm32_vref.c)
```

knthm replied:
It's already guarded in `drivers/sensor/CMakeLists.txt`, I think doing that here would be redundant.

GeorgeCGV commented on drivers/sensor/stm32_vref/stm32_vref.c:
```diff
@@ -0,0 +1,145 @@
+#endif /* CONFIG_SOC_SERIES_STM32H5X */
+
+LOG_MODULE_REGISTER(stm32_vref, CONFIG_SENSOR_LOG_LEVEL);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vref)
+#define DT_DRV_COMPAT st_stm32_vref
+#else
+#error "No compatible devicetree node found"
+#endif
```
Shouldn't be required. The file won't be built when `CONFIG_STM32_VREF` is not enabled and it depends on `DT_HAS_ST_STM32_VREF_ENABLED`.

GeorgeCGV commented on drivers/sensor/stm32_vref/stm32_vref.c:
```diff
@@ -0,0 +1,145 @@
+#endif /* CONFIG_SOC_SERIES_STM32H5X */
+
+LOG_MODULE_REGISTER(stm32_vref, CONFIG_SENSOR_LOG_LEVEL);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vref)
+#define DT_DRV_COMPAT st_stm32_vref
+#else
+#error "No compatible devicetree node found"
+#endif
```
Shouldn't be required. The file won't be built when `CONFIG_STM32_VREF` is not enabled and it depends on `DT_HAS_ST_STM32_VREF_ENABLED`.

knthm replied:
It's a leftover from the other `stm32_*`drivers, however you have a good point.

Will remove the preprocessor conditional for succinctness.

knthm commented:
- Remove superfluous `DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vref)` guard condition

gautierg-st commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -358,6 +359,13 @@
 		status = "disabled";
 	};
+
+	vbat: vbat {
```
Since STM32F0x0 don't have Vbat channel, you should add this node in stm32f031.dtsi. All non-F0x0 dtsi include this file. 
This would remove the need to delete the node for F030 and F070.
This is also true for the vbat-channel property.

knthm replied:
Oh I didn't see that, good catch, thanks! I'll amend the commit.

gautierg-st reviewed (changes requested):
One comment on STM32F0 dtsi, otherwise looks good

gautierg-st reviewed (dismissed):
LGTM. Thanks for this work

MaureenHelm commented on samples/sensor/soc_voltage/README.rst:
```diff
@@ -0,0 +1,30 @@
+.. _stm32_vref_sensor:
```
Need to update this anchor

knthm replied:
Good catch! Missed the README with my replace regex it seems when updating the sample.

MaureenHelm commented on samples/sensor/soc_voltage/README.rst:
```diff
@@ -0,0 +1,30 @@
+The tail ``0`` is the sensor number.  This sample support up to 16 sensors.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/sensor/stm32_vref_sensor
```
Need to fix the app path

knthm replied:
Same as above, thanks for the in-depth look. :)

MaureenHelm reviewed (changes requested):
Two small changes needed, but otherwise looks great!

knthm commented:
- Rebase on main
- Fix sample README references to old `stm32_vref_sensor` name
